kong/tools/uri.lua

```diff
@@ -12,52 +9,108 @@ local table_concat = table.concat
 local ngx_re_find = ngx.re.find
 local ngx_re_gsub = ngx.re.gsub
 
+local table_new = table.new
+
+
+-- Charset:
+--   reserved = "!" / "*" / "'" / "(" / ")" / ";" / ":" / "@" / "&" / "=" / "+" / "$" / "," / "/" / "?" / "%" / "#" / "[" / "]"
+--   unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
+--   other: * (meaning any char that is not mentioned above)
+
+-- Reserved characters have special meaning in URI. Encoding/decoding it affects the semantics of the URI;
+-- Unreserved characters are safe to use as part of HTTP message without encoding;
+-- Other characters has not special meaning but may be not safe to use as part of HTTP message without encoding;
+
+-- We should not unescape or escape reserved characters;
+-- We should unescape but not escape unreserved characters;
+-- We choose to unescape when processing and escape when forwarding for other characters
+
+local RESERVED_CHARS = "!*'();:@&=+$,/?%#[]"
+
+local HYPHEN_BYTE = string_byte('-')
+local DOT_BYTE = string_byte('.')
+local UNDERSCORE_BYTE = string_byte('_')
+local TILDE_BYTE = string_byte('~')
+local CAP_A_BYTE = string_byte('A')
+local CAP_Z_BYTE = string_byte('Z')
+local A_BYTE = string_byte('a')
+local Z_BYTE = string_byte('z')
+local ZERO_BYTE = string_byte('0')
+local NINE_BYTE = string_byte('9')
+
+local CHAR_RESERVED = true
+local CHAR_UNRESERVED = false
+local CHAR_OTHERS = nil -- luacheck: ignore
+
+local char_urlencode_type = table_new(256, 0) do
```

Comment:
```suggestion
local chars_to_decode = table_new(256, 0)
do
```

Is probably better, because when the value is `true`, character should be decoded.